### PR TITLE
[v1.1 Workflow] Added Distiller into the DE-NAS BERT

### PR DESCRIPTION
**Pull request by tianyil1**
1. Implemented the distiller as an option in the DE-NAS BERT training workflow
2. Added the supernet loading features for using distiller teacher model
3. Added CI/CD for the NLP workflow
4. Added Readme in DE-NAS NLP folder for using NLP workflow

This PR is tested well in the ICX env, where distiller can help improve the DE-NAS BERT model F1 score from 84.79 to 87.37 with the same DE-NAS searched structure.

### CODE REVIEW

File: e2eAIOK/DeNas/nlp/bert_trainer.py [hunk 1]
```diff
@@ -35,6 +39,15 @@ def _pre_process(self):
             custom_ops_thop = customer_ops_map_thop()
             macs_thop, _ = profile(self.model, inputs=(inputs,), custom_ops=custom_ops_thop)
             logging.info("(THOP) MACs: %.2f" % (macs_thop/(1000**3)))
+        if self.cfg.is_tl:
```
Comment (xuechendi) [inline e2eAIOK/DeNas/nlp/bert_trainer.py:42]: this argument name is hard to relate to is_transferlearning, @tianyil1 , can you come up with a more easy to understand name so user will recognize what this argument is used for.

File: e2eAIOK/DeNas/nlp/bert_trainer.py [hunk 2]
```diff
@@ -35,6 +39,15 @@ def _pre_process(self):
...
+            self.teacher_model = ModelBuilderNLPDeNas(self.cfg)._init_extra_model(self.cfg.teacher_model, self.cfg.teacher_model_structure)
+            self.teacher_distiller = kd.KD(pretrained_model=self.teacher_model, use_saved_logits=True)
+            self.logger.info("Successfully load teacher model!")
+            if self.cfg.is_saving_logits:
+                self.teacher_distiller.prepare_logits(self.train_dataloader, epochs=int(self.cfg.train_epochs))
+                self.logger.info("Successfully save teacher model logits!")
+                sys.exit()
```
Comment (xuechendi) [inline e2eAIOK/DeNas/nlp/bert_trainer.py:49]: why do sys.exit() after saving logits? does that mean saving logits will definitely happened at the end of the process.

File: e2eAIOK/DeNas/nlp/bert_trainer.py [hunk 3]
```diff
@@ -60,14 +73,20 @@ def _post_process(self):
...
+        if hasattr(self.train_dataloader.dataset, "set_epoch"):
+            self.train_dataloader.dataset.set_epoch(epoch)
 
         for step, batch in enumerate(tqdm(self.train_dataloader, desc="Iteration", ascii=True)):
             self.model.train()
             inputs, targets = batch
             outputs = self.model(inputs)
-            loss = self.criterion(outputs, targets)
+            if self.cfg.is_tl:
+                loss = self.model.loss(outputs, targets)
```
Comment (xuechendi) [inline e2eAIOK/DeNas/nlp/bert_trainer.py:87]: any chance we can merge this two option? looks to me that using 'model.loss' can also means using 'criterion' ? so we don't need to add a switch here?

File: e2eAIOK/DeNas/nlp/bert_trainer.py [hunk 4]
```diff
@@ -35,6 +39,15 @@ def _pre_process(self):
             custom_ops_thop = customer_ops_map_thop()
             macs_thop, _ = profile(self.model, inputs=(inputs,), custom_ops=custom_ops_thop)
             logging.info("(THOP) MACs: %.2f" % (macs_thop/(1000**3)))
+        if self.cfg.is_tl:
+            self.teacher_model = ModelBuilderNLPDeNas(self.cfg)._init_extra_model(self.cfg.teacher_model, self.cfg.teacher_model_structure)
```
Comment (xuechendi) [inline e2eAIOK/DeNas/nlp/bert_trainer.py:43]: I realized that this model can only be one download from huggingface by reading '_init_extra_model', in that case, can we add some pre-check or inform users with model-names not supported in huggingface? Or the teacher model is actually has a limited option list?

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 1 above)
Comment (tianyil1) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:42]: Got it. I will change the name to "is_transferlearning".

**Comment by xuechendi** [comment]
@tianyil1 , please also add UT or point me the existing test to test both w/wo is_tl enabled.

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 2 above)
Comment (tianyil1) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:49]: Yes, it means the saving-logits has saved the needed logits in the function "prepare_logits", and no need for the next steps, so call the "sys.exit()".

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 3 above)
Comment (tianyil1) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:87]: Got it, I will refine this issue. Thx.

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 1 above)
Comment (xuechendi) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:42]: @tianyil1 , is_transferlearning is just I trying to understand 'is_tl', after reading your codes, I am thinking since you also used self.cfg.teacher_model, how about simply using if 'self.cfg.teacher_model' is none or sth to indicate 'is_tl'?

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 4 above)
Comment (tianyil1) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:43]: Yes, the teacher model is currently a limited option list.

File: e2eAIOK/DeNas/nlp/bert_trainer.py (hunk 2 above)
Comment (xuechendi) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:49]: @tianyil1 , so can I understand as 'saving logits' and 'transferrable_model.make_transferrable_with_knowledge_distillation' is a if/else ? If user want to save logits, they may not to do transfer learning?

Comment (xuechendi) [reply e2eAIOK/DeNas/nlp/bert_trainer.py:49]: adhoc meeting with @tianyil1 to clarify this issue, create an issue to fix later https://github.com/intel/e2eAIOK/issues/168

**Comment by tianyil1** [comment]
All the above comments are fixed, please help check them. @xuechendi

**Review by xuechendi — APPROVED** [review]
Approved
